Add tutorials about karmada-search (#257)
karmada-io/website: docs/tutorials/karmada-search.md

title: Use Karmada-search to experience multi-cloud search
---

This guide will cover:

* Install `karmada-search` component in the Karmada control plane.
* Cache `Deployment` resources across multiple clusters.
* Use `OpenSearch` to graphically retrieve Kubernetes resources.

## Prerequisites

Before installing `karmada-search`, you must install Karmada control plane first. To start up Karmada, you can refer to [here](../installation/installation.md).
If you just want to try Karmada, we recommend building a development environment by `hack/local-up-karmada.sh`.

```shell
git clone https://github.com/karmada-io/karmada
cd karmada
hack/local-up-karmada.sh
```

## Install karmada-search

If you use `hack/local-up-karmada.sh`, `karmada-search` has been already installed.

> If you install Karmada by Helm, you can choose any of the following ways to install:

Reviewer: ```suggestion
If you install Karmada by Helm, you can choose any of the following ways to install:
```
```suggestion
If you install Karmada by Helm, you can choose any of the following ways to install `karmada-search`:
```

* Install `karmada-search` together in `host` mode
```shell
helm upgrade --install karmada -n karmada-system --create-namespace --dependency-update \
     --cleanup-on-fail ./charts/karmada \
     --set components={"search"}
```

* Install `karmada-search` separately in `component` mode

Edit values.yaml for karmada-search
```yaml
installMode: "component"
components: [
  "search"
]
...
```

Execute command (switch to the `root` directory of the repo, and sets the current-context in a kubeconfig file)
```shell
kubectl config use-context host
helm install karmada -n karmada-system ./charts/karmada
```

In addition, `karmadactl` supports one-click installation for `karmada-search`.
```shell
karmadactl addons enable karmada-search
```
For more details, you can refer to [karmadactl addons instruction](https://karmada.io/docs/next/reference/karmadactl/karmadactl-commands/karmadactl_addons/).

> ## Cache Deployment resources across multi-clusters

Reviewer: ```suggestion
## Cache Deployment resources across clusters
```

In the following steps, we are going to cache `Deployment` resources across member clusters.
Now we have propagated a nginx Deployment to member1 and member2 according to [the example](../get-started/nginx-example.md).

### 1. Create ResourceRegistry that will cache deployment across target clusters

```yaml
### deployment-search.yaml
apiVersion: search.karmada.io/v1alpha1
kind: ResourceRegistry
metadata:
  name: deployment-search
spec:
  targetCluster:
    clusterNames:
      - member1
      - member2
  resourceSelectors:
    - apiVersion: apps/v1
      kind: Deployment
```

```shell
kubectl --kubeconfig $HOME/.kube/karmada.config --context karmada-apiserver create -f deployment-search.yaml
```

> ### 2. Test By Kubernetes API

Reviewer: ```suggestion
### 2. Test using Kubernetes API
```

```shell
./hack/deploy-karmada-opensearch.sh $HOME/.kube/karmada.config karmada-host
```

Verify the installation:

```shell
kubectl --kubeconfig $HOME/.kube/karmada.config --context karmada-host get po -A
```

The output is similar to:

```
NAMESPACE            NAME                                                   READY   STATUS    RESTARTS   AGE
karmada-system       karmada-opensearch-77454fbcf5-7rpvz                    1/1     Running   0          155m
karmada-system       karmada-opensearch-dashboards-596bf4d9dd-n9429         1/1     Running   0          156m
...
```

### 2. Update ResourceRegistry with backendStore

```yaml
### deployment-search.yaml
apiVersion: search.karmada.io/v1alpha1
kind: ResourceRegistry
metadata:
  name: deployment-search
spec:
  backendStore:
    openSearch:
      addresses:
        - http://karmada-opensearch.karmada-system.svc:9200
  targetCluster:
    clusterNames:
      - member1
      - member2
  resourceSelectors:
    - apiVersion: apps/v1
      kind: Deployment
```

```shell
kubectl --kubeconfig $HOME/.kube/karmada.config --context karmada-apiserver create -f deployment-search.yaml
```

### 3. Expose the service of dashboard

> You need to expose the web service to the host port so that you can visit the dashboard by http.

Reviewer: ```suggestion
You need to expose the web service to the host port so that you can visit the dashboard by HTTP requrests.
```

```shell
kubectl --kubeconfig $HOME/.kube/karmada.config --context karmada-host port-forward svc/karmada-opensearch-dashboards 5601:5601 -nkarmada-system --address=0.0.0.0
```

### 4. Visit the dashboard

Visit OpenSearch dashboard(http://NodeIP:5601):

![opensearch](../resources/tutorials/opensearch.png)

> Now data of `Deployment` across member1 and member2 have uploaded to `OpenSearch`. 

Reviewer: ```suggestion
Now data of `Deployment` across member1 and member2 has been uploaded to `OpenSearch`. 
```

> You can experience multi-cloud search yourself by leveraging the search engine.

Reviewer: ```suggestion
You can try multi-cloud search yourself by leveraging the search engine.
```
